apisix/plugins/chaitin-waf.lua

```diff
@@ -0,0 +1,369 @@
+                -- request body size, in KB, integer, default 1MB (1024KB)
+                req_body_size = {
+                    type = "integer",
+                    default = 1024
+                },
+                -- maximum concurrent idle connections to
+                -- the SafeLine WAF detection service, integer, default 256
+                keepalive_size = {
+                    type = "integer",
+                    default = 256
+                },
+                -- idle connection timeout, in milliseconds, integer, default 60s (60000ms)
+                keepalive_timeout = {
+                    type = "integer",
+                    default = 60000
+                },
+                -- remote address from ngx.var.VARIABLE, string
+                remote_addr = {
+                    type = "string",
+                    default = "http_x_forwarded_for: 1",
+                }
+            },
+            default = {},
+        },
+    },
+    required = { "nodes" },
+}
+
+local global_server_picker
+
+local _M = {
+    version = 0.1,
+    priority = 2700,
+    name = plugin_name,
+    schema = plugin_schema,
+    metadata_schema = metadata_schema
+}
+
+function _M.check_schema(conf, schema_type)
+    if schema_type == core.schema.TYPE_METADATA then
+        return core.schema.check(metadata_schema, conf)
+    end
+
+    local ok, err = core.schema.check(plugin_schema, conf)
+
+    if not ok then
+        return false, err
+    end
+
+    if conf.match then
+        for _, m in ipairs(conf.match) do
+            local ok, err = expr.new(m.vars)
+            if not ok then
+                return false, "failed to validate the 'vars' expression: " .. err
+            end
+        end
+    end
+
+    return true
+end
+
+local function get_healthy_chaitin_server_nodes(metadata, checker)
+    local nodes = metadata.nodes
+    local new_nodes = core.table.new(0, #nodes)
+
+    for i = 1, #nodes do
+        local host, port = nodes[i].host, nodes[i].port
+        new_nodes[host .. ":" .. tostring(port)] = 1
+    end
+    return new_nodes
+end
+
+local function get_chaitin_server(metadata, ctx)
+    if not global_server_picker or global_server_picker.upstream ~= metadata.value.nodes then
+        local up_nodes = get_healthy_chaitin_server_nodes(metadata.value)
+        if core.table.nkeys(up_nodes) == 0 then
+            return nil, nil, "no healthy nodes"
+        end
+        core.log.info("chaitin-waf nodes: ", core.json.delay_encode(up_nodes))
+
+        global_server_picker = rr_balancer.new(up_nodes, metadata.value.nodes)
+    end
+
+    local server = global_server_picker.get(ctx)
+    local host, port, err = core.utils.parse_addr(server)
+    if err then
+        return nil, nil, err
+    end
+    return host, port, nil
+end
+
+local function check_match(conf, ctx)
+    local match_passed = true
+
+    if conf.match then
+        for _, match in ipairs(conf.match) do
+            local exp, err = expr.new(match.vars)
```

Comment:
Can we cache the result of expression compilation？

Reply:
We will do it after this pr

Reply:
Add TODO pls

Reply:
Done